### PR TITLE
Support unix domain sockets in the http client

### PR DESCRIPTION
**Pull request by seanyoung**
There are no standard for encoding a unix socket in an url. nginx uses:

http://unix:/path/to/unix/socket:/httppath

The second colon is needed to delimit where the unix path ends and where
the rest of the url continues.

### CODE REVIEW

File: http.c [hunk 1]
```diff
@@ -2316,6 +2307,59 @@ evhttp_connection_base_bufferevent_new(struct event_base *base, struct evdns_bas
 	return (NULL);
 }
 
+struct evhttp_connection *
+evhttp_connection_unixsocket_new(struct event_base *base, struct bufferevent* bev, const char *unixsocket)
```
Comment (azat) [inline http.c]: I think this should not exist for windows, i.e. the whole function should be guarded with ifdef

File: http-internal.h [hunk 2]
```diff
@@ -64,6 +64,9 @@ struct evhttp_connection {
 	char *bind_address;		/* address to use for binding the src */
 	unsigned short bind_port;	/* local port for binding the src */
 
+#ifndef _WIN32
+	char *unixsocket;
+#endif
```
Comment (azat) [inline http-internal.h]: put it after `port` please

File: http.c (hunk 1 above)
Comment (azat) [reply http.c]: and I don't like the namings, maybe it can be done in the same manner as `evhttp_connection_base_bufferevent_new`, i.e. `evhttp_connection_base_bufferevent_unix_new` (not sure though, since this is pretty long)

File: http.c [hunk 3]
```diff
@@ -2489,7 +2533,20 @@ evhttp_connection_connect_(struct evhttp_connection *evcon)
 			socklen = sizeof(struct sockaddr_in6);
 		}
 		ret = bufferevent_socket_connect(evcon->bufev, sa, socklen);
-	} else {
+	}
+#ifndef _WIN32
+	else if (evcon->unixsocket) {
+		struct sockaddr_un sockaddr;
+		sockaddr.sun_family = AF_UNIX;
+		if (strlen(evcon->unixsocket) >= sizeof(sockaddr.sun_path)) {
```
Comment (azat) [inline http.c]: this check should be in `evhttp_connection_unixsocket_new`

File: http.c [hunk 4]
```diff
@@ -4036,27 +4093,48 @@ evhttp_get_request_connection(
...
-	char *hostname = NULL, *portname = NULL;
 	struct bufferevent* bev = NULL;
 
-	name_from_addr(sa, salen, &hostname, &portname);
-	if (hostname == NULL || portname == NULL) {
-		if (hostname) mm_free(hostname);
-		if (portname) mm_free(portname);
-		return (NULL);
+#ifndef _WIN32
+	if (sa->sa_family == AF_UNIX) {
```
Comment (azat) [inline http.c]: this is not client functionality, so should be moved into a separate patch

File: http.c [hunk 5]
```diff
@@ -4083,7 +4161,7 @@ evhttp_associate_new_request_with_connection(struct evhttp_connection *evcon)
 	if ((req = evhttp_request_new(evhttp_handle_request, http)) == NULL)
 		return (-1);
 
-	if ((req->remote_host = mm_strdup(evcon->address)) == NULL) {
+	if ((req->remote_host = mm_strdup(evcon->address ?: "localhost")) == NULL) {
```
Comment (azat) [inline http.c]: AFAIR visualstudio do not like this syntax (sadly, but, true, 2a4bf294704034f6a7a962931d3028fe6c0b0ac8)

Comment (azat) [reply http.c]: And what the purpose of the default to localhost? Some issues in code for unixsocket, but this will hide some possible issues, no? (since you silently fallbackend to localhost)

File: http.c [hunk 6]
```diff
@@ -4277,6 +4355,9 @@ struct evhttp_uri {
 	char *userinfo; /* userinfo (typically username:pass), or NULL */
 	char *host; /* hostname, IP address, or NULL */
 	int port; /* port, or zero */
+#ifndef _WIN32
+	char *unixsocket; /* unix domain socket or NULL */
+#endif
```
Comment (azat) [inline http.c]: all changes to `evhttp_uri` should be in the separate patch too

File: http.c [hunk 7]
```diff
@@ -4787,6 +4891,15 @@ evhttp_uri_get_host(const struct evhttp_uri *uri)
 {
 	return uri->host;
 }
+const char *
+evhttp_uri_get_unixsocket(const struct evhttp_uri *uri)
```
Comment (azat) [inline http.c]: function should not exist for windows too

File: http.c (hunk 1 above)
Comment (seanyoung) [reply http.c]: Done and Renamed to evhttp_connection_base_bufferevent_unix_new

File: http.c (hunk 3 above)
Comment (seanyoung) [reply http.c]: done

File: http.c (hunk 5 above)
Comment (seanyoung) [reply http.c]: removed

File: http-internal.h (hunk 2 above)
Comment (seanyoung) [reply http-internal.h:66]: done

**Comment by seanyoung** [comment]
Please let me know if you want to split the commit when you come round to reviewing it in 2025.

File: test/regress_http.c [hunk 8]
```diff
@@ -2844,9 +2932,8 @@ http_parse_query_str_flags_test(void *ptr)
 static void
 http_parse_uri_test(void *ptr)
 {
-	const int nonconform = (ptr != NULL);
-	const unsigned parse_flags =
-	    nonconform ? EVHTTP_URI_NONCONFORMANT : 0;
```
Comment (azat) [inline test/regress_http.c:2849]: Loose `EVHTTP_URI_NONCONFORMANT`, the problem that tests does not found, because the test was wrong

File: test/regress_http.c [hunk 9]
```diff
@@ -2220,6 +2219,95 @@ http_dispatcher_test(void *arg)
...
+	struct evhttp_connection *evcon = NULL;
+	struct evhttp_request *req;
+
+	struct evhttp *myhttp = evhttp_new(data->base);
+
+	tt_assert(!evhttp_bind_unixsocket(myhttp, "foo"));
+
+	evhttp_set_cb(myhttp, "/", http_dispatcher_cb, data->base);
+
+	uri = evhttp_uri_parse_with_flags("http://unix:./foo:/?arg=val", EVHTTP_URI_UNIX_SOCKET);
```
Comment (azat) [inline test/regress_http.c:2282]: Static file - ok for now

File: http.c [hunk 10]
```diff
@@ -4466,24 +4522,46 @@ evhttp_get_request_connection(
...
-	if (hostname == NULL || portname == NULL) {
-		if (hostname) mm_free(hostname);
-		if (portname) mm_free(portname);
-		return (NULL);
+#ifndef _WIN32
+	if (sa->sa_family == AF_UNIX) {
+		struct sockaddr_un *sockaddr = (struct sockaddr_un *)sa;
+
+		event_debug(("%s: new request from unix socket %s on "
+			EV_SOCK_FMT"\n", __func__, sockaddr->sun_path,
```
Comment (azat) [inline http.c:4530]: This will not include the socket path anyway

File: http.c [hunk 11]
```diff
@@ -4910,6 +4993,18 @@ parse_authority(struct evhttp_uri *uri, char *s, char *eos, unsigned *flags)
...
 	}
+
+#ifndef _WIN32
+	if (*flags & EVHTTP_URI_UNIX_SOCKET && !strncmp(cp, "unix:", 5)) {
+		char *e = strchr(cp + 5, ':');
+		if (e) {
+			*e = '\0';
+			uri->unixsocket = mm_strdup(cp + 5);
+			return 0;
+		}
```
Comment (azat) [inline http.c:5004]: Let's return an error

**Review by azat — APPROVED** [review]
Thanks!

I've addressed my comments, but I can't push to your topic branch, will apply the patch to upstream w/o updating PR then, and close PR after (so it will not have merged icon).

**Comment by azat** [comment]
Applied. Upstream commit is f446229b2206691eae7ec892e68b2a0d6ef61306.
